File: backend/vendor-service.js
const { pool } = require('./db');

class VendorService {
  // ── CREATE VENDOR ────────────────────────────────────────────────────

  static async createVendor(vendorData) {
    try {
      const {
        name,
        trade,
        phone,
        email,
        address,
        city,
        state,
        zip,
        website,
        insurance_status,
        insurance_expiry
      } = vendorData;

      const result = await pool.query(
        `INSERT INTO vendors (name, trade, phone, email, address, city, state, zip, website, insurance_status, insurance_expiry)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
         RETURNING *`,
        [name, trade, phone, email, address, city, state, zip, website, insurance_status, insurance_expiry]
      );

      console.log('✓ Vendor created:', result.rows[0].id);
      return { success: true, vendor: result.rows[0] };
    } catch (err) {
      console.error('Create vendor error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── GET VENDOR BY ID ─────────────────────────────────────────────────

  static async getVendor(vendorId) {
    try {
      const result = await pool.query(
        'SELECT * FROM vendors WHERE id = $1',
        [vendorId]
      );

      if (result.rows.length === 0) {
        return { success: false, error: 'Vendor not found' };
      }

      return { success: true, vendor: result.rows[0] };
    } catch (err) {
      console.error('Get vendor error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── SEARCH VENDORS ───────────────────────────────────────────────────

  static async searchVendors(filters) {
    try {
      let query = 'SELECT * FROM vendors WHERE 1=1';
      const params = [];
      let paramCount = 1;

      // Name or trade search
      if (filters.search && filters.search.trim()) {
        query += ` AND (LOWER(name) LIKE LOWER($${paramCount}) OR LOWER(trade) LIKE LOWER($${paramCount}))`;
        params.push(`%${filters.search}%`);
        paramCount++;
      }

      // Trade filter
      if (filters.trade && filters.trade.trim()) {
        query += ` AND LOWER(trade) = LOWER($${paramCount})`;
        params.push(filters.trade);
        paramCount++;
      }

      // City filter - only if explicitly set and not empty
      if (filters.city && filters.city.trim()) {
        query += ` AND LOWER(city) = LOWER($${paramCount})`;
        params.push(filters.city);
        paramCount++;
      }

      // Insurance status filter
      if (filters.insurance_status && filters.insurance_status.trim()) {
        query += ` AND insurance_status = $${paramCount}`;
        params.push(filters.insurance_status);
        paramCount++;
      }

      // Minimum rating filter
      if (filters.min_rating) {
        query += ` AND avg_rating >= $${paramCount}`;
        params.push(parseFloat(filters.min_rating));
        paramCount++;
      }

      // Sorting
      const sortMap = {
        'rating': 'avg_rating DESC',
        'name': 'name ASC',
        'newest': 'created_at DESC',
        'reviews': 'review_count DESC'
      };
      const sortBy = sortMap[filters.sort] || 'created_at DESC';
      query += ` ORDER BY ${sortBy}`;

      // Pagination
      const limit = parseInt(filters.limit) || 50;
      const offset = parseInt(filters.offset) || 0;
      query += ` LIMIT $${paramCount} OFFSET $${paramCount + 1}`;
      params.push(limit, offset);

      console.log('Search query:', query);
      console.log('Params:', params);

      const result = await pool.query(query, params);
      return { success: true, vendors: result.rows, total: result.rows.length };
    } catch (err) {
      console.error('Search vendors error:', err);
      return { success: false, error: err.message };
    }
  }
  
  // ── UPDATE VENDOR ────────────────────────────────────────────────────

  static async updateVendor(vendorId, updates) {
    try {
      const allowedFields = [
        'name', 'trade', 'phone', 'email', 'address', 'city', 'state', 'zip',
        'website', 'insurance_status', 'insurance_expiry'
      ];

      const updateFields = [];
      const updateValues = [];
      let paramCount = 1;

      for (const [key, value] of Object.entries(updates)) {
        if (allowedFields.includes(key) && value !== undefined) {
          updateFields.push(`${key} = $${paramCount}`);
          updateValues.push(value);
          paramCount++;
        }
      }

      if (updateFields.length === 0) {
        return { success: false, error: 'No valid fields to update' };
      }

      updateValues.push(vendorId);
      const query = `UPDATE vendors SET ${updateFields.join(', ')}, updated_at = NOW() WHERE id = $${paramCount} RETURNING *`;

      const result = await pool.query(query, updateValues);

      if (result.rows.length === 0) {
        return { success: false, error: 'Vendor not found' };
      }

      console.log('✓ Vendor updated:', vendorId);
      return { success: true, vendor: result.rows[0] };
    } catch (err) {
      console.error('Update vendor error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── ADD REVIEW ───────────────────────────────────────────────────────

  static async addReview(vendorId, userId, reviewData) {
    try {
      const { rating_reliability, rating_cost, rating_quality, rating_communication, rating_insurance, comment } = reviewData;

      // Check if user already reviewed this vendor
      const existingReview = await pool.query(
        'SELECT id FROM vendor_reviews WHERE vendor_id = $1 AND user_id = $2',
        [vendorId, userId]
      );

      if (existingReview.rows.length > 0) {
        return { success: false, error: 'User has already reviewed this vendor' };
      }

      const result = await pool.query(
        `INSERT INTO vendor_reviews (vendor_id, user_id, rating_reliability, rating_cost, rating_quality, rating_communication, rating_insurance, comment)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
         RETURNING *`,
        [vendorId, userId, rating_reliability, rating_cost, rating_quality, rating_communication, rating_insurance, comment]
      );

      // Update vendor average rating
      await this.updateVendorRating(vendorId);

      console.log('✓ Review added for vendor:', vendorId);
      return { success: true, review: result.rows[0] };
    } catch (err) {
      console.error('Add review error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── UPDATE VENDOR RATING ─────────────────────────────────────────────

  static async updateVendorRating(vendorId) {
    try {
      const result = await pool.query(
        `UPDATE vendors 
         SET avg_rating = (
           SELECT AVG((rating_reliability + rating_cost + rating_quality + rating_communication + rating_insurance) / 5)
           FROM vendor_reviews WHERE vendor_id = $1
         ),
         review_count = (SELECT COUNT(*) FROM vendor_reviews WHERE vendor_id = $1)
         WHERE id = $1
         RETURNING avg_rating, review_count`,
        [vendorId]
      );

      if (result.rows.length > 0) {
        console.log(`✓ Vendor ${vendorId} rating updated:`, result.rows[0]);
      }
    } catch (err) {
      console.error('Update rating error:', err);
    }
  }

  // ── GET VENDOR REVIEWS ───────────────────────────────────────────────

  static async getVendorReviews(vendorId, limit = 10, offset = 0) {
    try {
      const result = await pool.query(
        `SELECT vr.*, u.name as reviewer_name 
         FROM vendor_reviews vr
         LEFT JOIN users u ON vr.user_id = u.id
         WHERE vr.vendor_id = $1
         ORDER BY vr.created_at DESC
         LIMIT $2 OFFSET $3`,
        [vendorId, limit, offset]
      );

      return { success: true, reviews: result.rows };
    } catch (err) {
      console.error('Get reviews error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── BULK IMPORT FROM CSV ─────────────────────────────────────────────

  static async bulkImportVendors(vendorsArray, userId) {
    try {
      const imported = [];
      const failed = [];

      for (let i = 0; i < vendorsArray.length; i++) {
        const vendor = vendorsArray[i];

        // Validate required fields
        if (!vendor.name || !vendor.trade) {
          failed.push({ row: i + 1, error: 'Missing name or trade' });
          continue;
        }

        try {
          const result = await pool.query(
            `INSERT INTO vendors (name, trade, phone, email, address, city, state, zip, website, insurance_status)
             VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
             ON CONFLICT (name, city) DO NOTHING
             RETURNING id`,
            [
              vendor.name,
              vendor.trade,
              vendor.phone || null,
              vendor.email || null,
              vendor.address || null,
              vendor.city || null,
              vendor.state || null,
              vendor.zip || null,
              vendor.website || null,
              vendor.insurance_status || 'not_verified'
            ]
          );

          if (result.rows.length > 0) {
            imported.push({ id: result.rows[0].id, name: vendor.name });

            // Log import
            await pool.query(
              `INSERT INTO vendor_imports (user_id, vendor_id, import_date, source)
               VALUES ($1, $2, NOW(), 'csv')`,
              [userId, result.rows[0].id]
            );
          }
        } catch (rowErr) {
          failed.push({ row: i + 1, error: rowErr.message });
        }
      }

      console.log(`✓ Bulk import complete: ${imported.length} imported, ${failed.length} failed`);
      return { success: true, imported, failed };
    } catch (err) {
      console.error('Bulk import error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── CLAIM VENDOR ACCOUNT ─────────────────────────────────────────────

  static async claimVendorAccount(vendorId, vendorEmail, vendorPassword) {
    try {
      // Check if already claimed
      const existing = await pool.query(
        'SELECT id FROM vendor_accounts WHERE vendor_id = $1',
        [vendorId]
      );

      if (existing.rows.length > 0) {
        return { success: false, error: 'Vendor account already claimed' };
      }

      const bcrypt = require('bcryptjs');
      const passwordHash = await bcrypt.hash(vendorPassword, 12);

      const result = await pool.query(
        `INSERT INTO vendor_accounts (vendor_id, email, password_hash, is_verified)
         VALUES ($1, $2, $3, false)
         RETURNING id`,
        [vendorId, vendorEmail, passwordHash]
      );

      console.log('✓ Vendor account created:', vendorId);
      return { success: true, account: result.rows[0] };
    } catch (err) {
      console.error('Claim account error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── TRACK VENDOR USAGE ───────────────────────────────────────────────

  static async trackVendorUsage(userId, vendorId, action) {
    try {
      await pool.query(
        `INSERT INTO vendor_usage_tracking (user_id, vendor_id, action_type)
         VALUES ($1, $2, $3)`,
        [userId, vendorId, action]
      );
    } catch (err) {
      console.error('Track usage error:', err);
    }
  }
}

module.exports = VendorService;

File: fastapi_backend/services/vendor_service.py
import bcrypt
from db import get_pool


ALLOWED_VENDOR_FIELDS = [
    "name", "trade", "phone", "email", "address", "city", "state", "zip",
    "website", "insurance_status", "insurance_expiry"
]


async def create_vendor(data: dict, project_id: int | None = None) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        try:
            async with conn.transaction():
                row = await conn.fetchrow(
                    """INSERT INTO vendors (name, trade, phone, email, address, city, state, zip, website, insurance_status, insurance_expiry)
                       VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11) RETURNING *""",
                    data.get("name"), data.get("trade"), data.get("phone"), data.get("email"),
                    data.get("address"), data.get("city"), data.get("state"), data.get("zip"),
                    data.get("website"), data.get("insurance_status"), data.get("insurance_expiry"),
                )
                # Link to the project the user was scoped to when they added
                # this vendor — otherwise it's created but invisible the
                # moment a project filter is applied (search_vendors inner-
                # joins project_vendors), which looked like "vendors don't
                # load after I add one."
                if project_id:
                    await conn.execute(
                        """INSERT INTO project_vendors (project_id, vendor_id) VALUES ($1,$2)
                           ON CONFLICT (project_id, vendor_id) DO NOTHING""",
                        project_id, row["id"],
                    )
            return {"success": True, "vendor": dict(row)}
        except Exception as e:
            return {"success": False, "error": str(e)}


async def get_vendor(vendor_id: int) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        row = await conn.fetchrow("SELECT * FROM vendors WHERE id = $1", vendor_id)
        if not row:
            return {"success": False, "error": "Vendor not found"}
        return {"success": True, "vendor": dict(row)}


async def search_vendors(filters: dict) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        select_clause = "SELECT v.*"
        from_clause = "FROM vendors v"
        params = []
        p = 1

        if filters.get("project_id"):
            select_clause = "SELECT DISTINCT v.*"
            from_clause += f" JOIN project_vendors pv ON pv.vendor_id = v.id AND pv.project_id = ${p}"
            params.append(int(filters["project_id"]))
            p += 1

        query = f"{select_clause} {from_clause} WHERE 1=1"

        if filters.get("search"):
            query += f" AND (LOWER(v.name) LIKE LOWER(${p}) OR LOWER(v.trade) LIKE LOWER(${p}))"
            params.append(f"%{filters['search']}%")
            p += 1
        if filters.get("trade"):
            query += f" AND LOWER(v.trade) = LOWER(${p})"
            params.append(filters["trade"])
            p += 1
        if filters.get("city"):
            query += f" AND LOWER(v.city) = LOWER(${p})"
            params.append(filters["city"])
            p += 1
        if filters.get("insurance_status"):
            query += f" AND v.insurance_status = ${p}"
            params.append(filters["insurance_status"])
            p += 1
        if filters.get("min_rating"):
            query += f" AND v.avg_rating >= ${p}"
            params.append(float(filters["min_rating"]))
            p += 1

        sort_map = {"rating": "v.avg_rating DESC", "name": "v.name ASC", "newest": "v.created_at DESC", "reviews": "v.review_count DESC"}
        query += f" ORDER BY {sort_map.get(filters.get('sort', 'newest'), 'v.created_at DESC')}"

        limit = int(filters.get("limit", 50))
        offset = int(filters.get("offset", 0))
        query += f" LIMIT ${p} OFFSET ${p+1}"
        params.extend([limit, offset])

        rows = await conn.fetch(query, *params)
        return {"success": True, "vendors": [dict(r) for r in rows], "total": len(rows)}


async def link_vendor_to_project(vendor_id: int, project_id: int) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        vendor = await conn.fetchval("SELECT id FROM vendors WHERE id = $1", vendor_id)
        if not vendor:
            return {"success": False, "error": "Vendor not found"}
        project = await conn.fetchval("SELECT id FROM projects WHERE id = $1", project_id)
        if not project:
            return {"success": False, "error": "Project not found"}
        row = await conn.fetchrow(
            """INSERT INTO project_vendors (project_id, vendor_id) VALUES ($1,$2)
               ON CONFLICT (project_id, vendor_id) DO NOTHING RETURNING *""",
            project_id, vendor_id,
        )
        if not row:
            row = await conn.fetchrow(
                "SELECT * FROM project_vendors WHERE project_id = $1 AND vendor_id = $2",
                project_id, vendor_id,
            )
        return {"success": True, "link": dict(row)}


async def update_vendor(vendor_id: int, updates: dict) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        fields, values = [], []
        p = 1
        for key, val in updates.items():
            if key in ALLOWED_VENDOR_FIELDS and val is not None:
                fields.append(f"{key} = ${p}")
                values.append(val)
                p += 1
        if not fields:
            return {"success": False, "error": "No valid fields to update"}
        values.append(vendor_id)
        row = await conn.fetchrow(
            f"UPDATE vendors SET {', '.join(fields)}, updated_at = NOW() WHERE id = ${p} RETURNING *", *values
        )
        if not row:
            return {"success": False, "error": "Vendor not found"}
        return {"success": True, "vendor": dict(row)}


async def _update_vendor_rating(conn, vendor_id: int):
    await conn.execute(
        """UPDATE vendors SET
           avg_rating = (SELECT AVG((rating_reliability + rating_cost + rating_quality + rating_communication + rating_insurance) / 5.0) FROM vendor_reviews WHERE vendor_id = $1),
           review_count = (SELECT COUNT(*) FROM vendor_reviews WHERE vendor_id = $1)
           WHERE id = $1""",
        vendor_id,
    )


async def add_review(vendor_id: int, user_id: int, review_data: dict) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        existing = await conn.fetchrow("SELECT id FROM vendor_reviews WHERE vendor_id = $1 AND user_id = $2", vendor_id, user_id)
        if existing:
            return {"success": False, "error": "User has already reviewed this vendor"}
        row = await conn.fetchrow(
            """INSERT INTO vendor_reviews (vendor_id, user_id, rating_reliability, rating_cost, rating_quality, rating_communication, rating_insurance, comment)
               VALUES ($1,$2,$3,$4,$5,$6,$7,$8) RETURNING *""",
            vendor_id, user_id,
            review_data.get("rating_reliability"), review_data.get("rating_cost"),
            review_data.get("rating_quality"), review_data.get("rating_communication"),
            review_data.get("rating_insurance"), review_data.get("comment"),
        )
        await _update_vendor_rating(conn, vendor_id)
        return {"success": True, "review": dict(row)}


async def get_vendor_reviews(vendor_id: int, limit: int = 10, offset: int = 0) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        rows = await conn.fetch(
            """SELECT vr.*, u.name as reviewer_name FROM vendor_reviews vr
               LEFT JOIN users u ON vr.user_id = u.id
               WHERE vr.vendor_id = $1 ORDER BY vr.created_at DESC LIMIT $2 OFFSET $3""",
            vendor_id, limit, offset,
        )
        return {"success": True, "reviews": [dict(r) for r in rows]}


async def bulk_import_vendors(vendors_array: list, user_id: int, project_id: int | None = None) -> dict:
    pool = await get_pool()
    imported, failed = [], []
    async with pool.acquire() as conn:
        for i, vendor in enumerate(vendors_array):
            if not vendor.get("name") or not vendor.get("trade"):
                failed.append({"row": i + 1, "error": "Missing name or trade"})
                continue
            try:
                row = await conn.fetchrow(
                    """INSERT INTO vendors (name, trade, phone, email, address, city, state, zip, website, insurance_status)
                       VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10) ON CONFLICT (name, city) DO NOTHING RETURNING id""",
                    vendor.get("name"), vendor.get("trade"), vendor.get("phone"),
                    vendor.get("email"), vendor.get("address"), vendor.get("city"),
                    vendor.get("state"), vendor.get("zip"), vendor.get("website"),
                    vendor.get("insurance_status", "not_verified"),
                )
                if row:
                    imported.append({"id": row["id"], "name": vendor["name"]})
                    await conn.execute(
                        "INSERT INTO vendor_imports (user_id, vendor_id, import_date, source) VALUES ($1,$2,NOW(),'csv')",
                        user_id, row["id"],
                    )
                    # Same project-linking gap as create_vendor — an
                    # imported vendor is invisible under a project filter
                    # until it's tied to that project.
                    if project_id:
                        await conn.execute(
                            """INSERT INTO project_vendors (project_id, vendor_id) VALUES ($1,$2)
                               ON CONFLICT (project_id, vendor_id) DO NOTHING""",
                            project_id, row["id"],
                        )
            except Exception as e:
                failed.append({"row": i + 1, "error": str(e)})
    return {"success": True, "imported": imported, "failed": failed}


async def claim_vendor_account(vendor_id: int, email: str, password: str) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        existing = await conn.fetchrow("SELECT id FROM vendor_accounts WHERE vendor_id = $1", vendor_id)
        if existing:
            return {"success": False, "error": "Vendor account already claimed"}
        password_hash = bcrypt.hashpw(password.encode(), bcrypt.gensalt(12)).decode()
        row = await conn.fetchrow(
            "INSERT INTO vendor_accounts (vendor_id, email, password_hash, is_verified) VALUES ($1,$2,$3,false) RETURNING id",
            vendor_id, email, password_hash,
        )
        return {"success": True, "account": dict(row)}
